doc/containers.md

```diff
@@ -317,6 +317,11 @@ the average of the limits will be used.
 Unix character device entries simply make the requested character device
 appear in the container's `/dev` and allow read/write operations to it.
 
+If they exist at the time of the LXD daemon start the "required" property is
+available for devices in the following directories: `/dev`, `/dev/dri`,
+`/dev/infiniband`, `/dev/input`, `/dev/mapper`, `/dev/net`, `/dev/pts`,
+`/dev/raw`, `/dev/thinkpad`, `/dev/scanners`, `/dev/usb`, `/dev/video`.
```

Comment:
This looks super hackish to me and hopefully we can do better than that :)

Our internal inotify handler should:
 - The parent directory of any path-based unix-block or unix-char device that has required=false.
 - If we notice a deletion event, then we should yank the device out of the container.
 - If the device doesn't exist in the container and we get a file creation event, then we pass it to the container.
 - If the parent of the device doesn't exist yet, then we find its closest existing parent and setup a watch on that, watching for the creation of the next directory, then setting up a watch on that, ... until we end up watching the path we want.

That last part is the main frustration with inotify, there's no nice way of dealing with recursive watches other than doing it yourself and you don't want to end up monitoring everything as you'd be wasting watches, so you need to carefully maintain a pool of watches, moving the watch further and further as paths get created.

Reply:
> This looks super hackish to me and hopefully we can do better than that :)

The first part of the prior spec:
> Maintain a central inotify handler for devices and make sure we subscribe to notifications for the parent directory of any unix device that we know of.

read like that.

> Our internal inotify handler should:
> - The parent directory of any path-based unix-block or unix-char device that has required=false.

Doing it already.

> - If we notice a deletion event, then we should yank the device out of the container.

Doing it already.

> - If the device doesn't exist in the container and we get a file creation event, then we pass it to the container.

Doing it already.

> - If the parent of the device doesn't exist yet, then we find its closest existing parent and setup a watch on that, watching for the creation of the next directory, then setting up a watch on that, ... until we end up watching the path we want.

> That last part is the main frustration with inotify, there's no nice way of dealing with recursive watches other than doing it yourself and you don't want to end up monitoring everything as you'd be wasting watches, so you need to carefully maintain a pool of watches, moving the watch further and further as paths get created.

One reason why I didn't do this is that it either requires watches to be per container or they are per path but reference counted.

Reply:
We likely have to reference count.

Reply:
But I haven't made up my mind yet.

Reply:
So I don't really like the reference counting idea because it seems unclean and error prone. For example, the container could simply exit behind our back and the corresponding value in the daemon struct is not decremented.

Reply:
For me, the real test for this feature will be:
 - Setup a container with a unix-char for /dev/serial/by-id/usb-FTDI_FT232R_USB_UART_A60337Y1-if00-port0 as /dev/ttyUSB0
 - Plug that USB serial device
 - Have /dev/ttyUSB0 available in the container
 - Unplug the device
 - It's gone from the container

The /dev/serial/by-id/... stuff is a bunch of symlinks generated from driver metadata and the type of path our users are pretty likely to use as those are consistent regardless of detection order (same as /dev/disk/by-* and some of the /dev/mapper devices).

The notification APIs in the kernel are all pretty crap to deal with that though.
An idea I had was to simply re-scan the paths we care about in /dev when the inode count of /dev changes, but I don't think any of the kernel notification APIs will give us an event for that...

Reply:
> The notification APIs in the kernel are all pretty crap to deal with that though.
An idea I had was to simply re-scan the paths we care about in /dev when the inode count of /dev changes, but I don't think any of the kernel notification APIs will give us an event for that...

Nope, there's no api to give us this info actually. The algorithm I thought about is essentially:

#### Handle non-existing intermediate directories
Consider a container wants to watch for an event on the device
"/mnt/a/b/my-device" but the intermediate directories "a" and "b" do not exist.
To handle this case we place a watch on the nearest existing ancestor
directory. In the example above we start by placing a watch on "/mnt". When we
observe a create event for the subdirectory "a" we place a watch on "/mnt/a".
When we observe a crete event for the subdirectory "b" we place a watch on
"/mnt/a/b".

- event "ev" is observed indicating an event for "a"
- check whether "a" refers to directory
- check whether any containers are waiting on the creation of devices in the
  subdirectory "a"
  - if so, add the subdirectroy "a" to the inotify targets
- check whether there are any containers waiting on the creation of devices in
  the ancestor directory "/mnt" of "a"
  - if so, keep "/mnt" as inotify target
  - if not, remove "/mnt" from inotify targets

- if a delete event on a directory is observed, check whether any existing
  containers still wait on events within that directory.
  - if so add the closest living ancestor of the deleted directory as inotify
    target

Reply:
Yep, sounds good. Also to prevent a race, you'll want to setup the new watch and then check that the thing you're waiting for didn't show up in the meantime.

Sounds like this will turn into another sub-package (lxd/inotify or something) which offers the interface we can use from LXD itself to monitor stuff (keeping things separate to make testing and re-use easier).